### PR TITLE
add discovery throttling

### PR DESCRIPTION
**Pull request by zzantozz**
The new Elasticsearch rest code seems to put more load on
Elasticsearch than the old code. Combined with heavy load, multiple
production nodes, and too-short cache TTLs, deploying new ingest code
and restarting is overloading Elasticsearch and affecting the entire
system, including other ingest nodes and query node performance.

This adds throttling to cap the number of discovery writes that will
be made per minute. Setting it lower reduces the impact on
Elasticsearch. There's both a global and per-tenant throttle.

The throttles are implemented here using Guava caches set to a
1-minute TTL. Local profiling of a similar cache suggests that the
cache itself uses approximately 70M of memory per million cache
entries, excluding the size of the entries.

### CODE REVIEW

**Review by samwar — APPROVED** [review]
The tests look sound, and the implementation does indeed appear to throttle the discovery input.

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/inputs/processors/DiscoveryWriter.java [hunk 1]
```diff
@@ -91,7 +138,11 @@ private static List<IMetric> condense(List<List<IMetric>> input) {
             }
 
             for (IMetric m : list) {
-                if (!LocatorCache.getInstance().isLocatorCurrentInDiscoveryLayer(m.getLocator())) {
+                boolean isAlreadySeen = LocatorCache.getInstance().isLocatorCurrentInDiscoveryLayer(m.getLocator());
+                boolean isThrottlingGlobally = newLocatorsThrottle.size() >= maxNewLocatorsPerMinute;
```
Comment (iWebi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/inputs/processors/DiscoveryWriter.java:142]: global throttle check can be done once for all metrics, isn't it?

**Review by iWebi — APPROVED** [review]
👍

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/inputs/processors/DiscoveryWriter.java (hunk 1 above)
Comment (zzantozz) [reply blueflood-core/src/main/java/com/rackspacecloud/blueflood/inputs/processors/DiscoveryWriter.java:142]: You're right. Good catch. Fixed in #913

**Pull request merged**